[SourceKit/CursorInfo] Mark dynamic calls

bnbarham opened the pull request:
Adds two new fields to the cursor info response:
  1. is_dynamic: whether a call is dynamic
  2. receivers: receivers of the call (USRs)

Users of the CursorInfo request can use "is_dynamic" to decide whether
to lookup overrides or not, and then the "receivers" as the starting
point of the lookup.

Resolves rdar://75385900

bnbarham committed:
[IDE] Refactor some call related functions out of Index.cpp

The cursor info request also needs to output whether a call is "dynamic"
or not, move the functions related to this out of Index.cpp and into
IDE/Utils.cpp.

Also cleanup the `TrailingExpr` handling in `CursorInfoResolver` - it
only needs the first expression.

bnbarham commented:
@swift-ci please test

bnbarham commented on lib/IDE/Utils.cpp:
```diff
@@ -1181,3 +1181,75 @@ std::pair<Type, ConcreteDeclRef> swift::ide::getReferencedDecl(Expr *expr) {
+    return false;
+
+  // super.method(), but is dynamic if the decl is marked as dynamic
+  if (Base->isSuperExpr() && !D->isDynamic())
```
This used to use isSuperRefExpr, which also checked for a LoadExpr (isSuperExpr does not). Removing it didn't cause any tests to fail... anyone know what case that would have been for?

akyrtzi replied:
I think we should dig into this a bit:
* Construct a test case to see what can happen here. Is the "injection" possible in pure Swift code or is this only for @objc classes?
* If we are going to mark a `super` call as dynamic we need to distinguish this case from other dynamic calls, otherwise this may get handled as the other dynamic calls that may go to override classes.

Without additional information to distinguish the cases I would argue it is better to treat it as "non-dynamic" since in the vast majority of cases it will do the right thing and go to the one function in the super class, instead of treating it as override methods may get called which is actually always wrong.

bnbarham replied:
It's only possible with @objc and dynamic. You're right that we probably want to distinguish this separately, including in the index. I'll leave this to another PR.

bnbarham commented on lib/IDE/Utils.cpp:
```diff
@@ -1181,3 +1181,75 @@ std::pair<Type, ConcreteDeclRef> swift::ide::getReferencedDecl(Expr *expr) {
+  if (!TyD)
+    return false;
+
+  if (isa<StructDecl>(TyD) || isa<EnumDecl>(TyD) || D->isFinal())
```
isFinal is new here

akyrtzi commented on tools/SourceKit/tools/sourcekitd/lib/API/Requests.cpp:
```diff
@@ -1834,6 +1834,8 @@ static void reportCursorInfo(const RequestResult<CursorInfoData> &Result,
       Parent.set(KeyUSR, ParentTy.USR);
     }
   }
+  if (Info.IsDynamic)
```
I think an important piece of this is also providing the USR of the receiver type. This can be used to determine at which point in the class hierarchy we should start looking for overrides.

bnbarham replied:
Added a "receivers" since protocol composition could make this multiple types (even though we currently don't handle that case).

bnbarham committed:
[SourceKit/CursorInfo] Mark dynamic calls

Adds two new fields to the cursor info response:
  1. is_dynamic: whether a call is dynamic
  2. receivers: receivers of the call (USRs)

Users of the CursorInfo request can use "is_dynamic" to decide whether
to lookup overrides or not, and then the "receivers" as the starting
point of the lookup.

Resolves rdar://75385900

bnbarham commented:
@swift-ci please test

bnbarham commented on lib/IDE/Utils.cpp:
```diff
@@ -1181,3 +1181,111 @@ std::pair<Type, ConcreteDeclRef> swift::ide::getReferencedDecl(Expr *expr) {
+    ReceiverTy = OpenedTy->getOpenedExistentialType();
+
+  if (auto TyD = ReceiverTy->getAnyNominal()) {
+    Types.push_back(TyD);
```
We currently add receivers for non-dynamic calls as well - do we want to keep that?

akyrtzi replied:
Let's keep it.

bnbarham commented on utils/gyb_sourcekit_support/UIDs.py:
```diff
@@ -38,6 +38,7 @@ def __init__(self, internal_name, external_name):
     KEY('Line', 'key.line'),
     KEY('Column', 'key.column'),
     KEY('ReceiverUSR', 'key.receiver_usr'),
+    KEY('Receivers', 'key.receivers'),
```
I wasn't sure whether to keep the "USR" part or not, some of the keys include it and some don't. Thoughts? (Eg. Overrides is not OverrideUSRs)

akyrtzi replied:
I prefer 'Receivers', and it'd be good not to just have it a flat list of USRs, but array of dictionaries, even if the only entry is a 'key.usr' in the beginning.
This will give us flexibility to extend it with more info in the future.

bnbarham replied:
Ah that's true, it is actually already a dictionary with key.usr so yep, adding "USRs" wouldn't make much sense!

bnbarham commented:
@swift-ci please test Linux platform

bnbarham commented:
@swift-ci please test Linux platform

bnbarham commented:
@benlangmuir going to merge but let me know if there's any changes you'd like me to make.
